File: streamflowSignatures/julia/src/storage.jl
"""
Average storage signature.

Catchment storage using simplified water balance (ignores ET).
"""

"""
    calculate_average_storage(df::DataFrame; min_years=10) -> Dict

Calculate average catchment storage trends.

Uses simplified water balance (dS = P - Q) without ET estimation.
Calculates storage amount corresponding to average discharge.

Parameters
----------
df : DataFrame
    Daily streamflow and climate data with columns: water_year, Q, PPT, dowy
min_years : Int
    Minimum years required

Returns
-------
Dict{String, Float64}
    Dictionary with avg_storage statistics (8 values)

Notes
-----
This calculation ignores evapotranspiration (ET), using only P - Q
for the water balance. This simplification may overestimate storage
in watersheds with significant ET losses.
"""
function calculate_average_storage(
    df::DataFrame;
    min_years::Int=10,
    trend_completeness::Union{Nothing, Float64}=nothing,
    decade_completeness::Union{Nothing, Float64}=nothing, min_values_for_stats::Union{Nothing, Int}=nothing,
    changepoint::Union{Nothing, NamedTuple}=nothing,
    collector::Union{Nothing, AnnualCollector}=nothing
)
    result = Dict{String, Float64}()

    # Check for required columns
    if !("PPT" in names(df)) || !("dowy" in names(df))
        return empty_stats("avg_storage")
    end

    if nrow(df) == 0
        return empty_stats("avg_storage")
    end

    years = unique(df.water_year)

    if length(years) < min_years
        return empty_stats("avg_storage")
    end

    annual_storage = Float64[]
    storage_years = Int[]

    for yr in years
        year_mask = coalesce.(df.water_year .== yr, false)
        if !any(year_mask)
            continue
        end
        year_df = copy(df[year_mask, :])

        # Convert to clean vectors
        Q_clean = coalesce_q(year_df.Q)
        P_clean = coalesce_q(year_df.PPT)

        # Skip year if any NaNs remain (preprocessor guarantees valid years have no NAs)
        Q_vals = copy(Q_clean)
        P_vals = copy(P_clean)
        if any(isnan, Q_vals) || any(isnan, P_vals)
            continue
        end

        # Sort by day of water year (handle Missing dowy)
        dowy_clean = [coalesce(d, 999) for d in year_df.dowy]
        perm = sortperm(dowy_clean)
        Q_vals = Q_vals[perm]
        P_vals = P_vals[perm]

        # Water balance: dS = P - Q
        dS = P_vals .- Q_vals

        # Cumulative storage (relative to start of year)
        S = cumsum(dS)

        # Mean Q for this year
        mean_Q = mean(Q_vals)

        # Sort by Q for interpolation
        sort_idx = sortperm(Q_vals)
        Q_sorted = Q_vals[sort_idx]
        S_sorted = S[sort_idx]

        # Remove NaN pairs defensively (preprocessed years should carry none)
        valid_mask = .!isnan.(Q_sorted) .& .!isnan.(S_sorted)
        Q_valid = Q_sorted[valid_mask]
        S_valid = S_sorted[valid_mask]

        if length(Q_valid) < 10
            continue
        end

        # Average S values at duplicate Q points to match R's approx(ties=mean).
        # Without this, picking a single S at tied Q points introduces noise
        # that corrupts trend statistics (slopes, p-values).
        Q_unique = unique(Q_valid)
        S_means = Float64[mean(S_valid[Q_valid .== q]) for q in Q_unique]

        if length(Q_unique) < 10
            continue
        end

        # Linear interpolation to find S at mean_Q (using de-duplicated arrays)
        idx_below = findlast(Q_unique .<= mean_Q)
        idx_above = findfirst(Q_unique .>= mean_Q)

        if idx_below === nothing || idx_above === nothing
            # mean_Q is outside the range
            if mean_Q < Q_unique[1]
                avg_storage = S_means[1]
            else
                avg_storage = S_means[end]
            end
        elseif idx_below == idx_above
            avg_storage = S_means[idx_below]
        else
            # Linear interpolation
            Q_lo, Q_hi = Q_unique[idx_below], Q_unique[idx_above]
            S_lo, S_hi = S_means[idx_below], S_means[idx_above]

            if abs(Q_hi - Q_lo) < 1e-10
                avg_storage = (S_lo + S_hi) / 2
            else
                t = (mean_Q - Q_lo) / (Q_hi - Q_lo)
                avg_storage = S_lo + t * (S_hi - S_lo)
            end
        end

        if !isnan(avg_storage)
            push!(annual_storage, avg_storage)
            push!(storage_years, yr)
        end
    end

    if length(annual_storage) < 5
        return empty_stats("avg_storage")
    end

    annual_df = DataFrame(
        water_year = storage_years,
        avg_storage = annual_storage
    )

    return generate_stats(annual_df; value_cols=["avg_storage"], trend_completeness=trend_completeness, decade_completeness=decade_completeness, min_values_for_stats=min_values_for_stats, changepoint=changepoint, collector=collector)
end

File: streamflowSignatures/python/streamflow_signatures/storage.py
"""
Average storage signature.

Calculates catchment storage using simplified water balance.
Requires precipitation (PPT) data.
"""

from typing import Dict, Optional
import numpy as np
import pandas as pd
from .stats import generate_stats
from .config import STORAGE_MIN_YEARS


def calculate_average_storage(
    streamflow_data: pd.DataFrame,
    min_years: int = STORAGE_MIN_YEARS,
    trend_completeness: Optional[float] = None,
    decade_completeness: Optional[float] = None,
    min_values_for_stats: Optional[int] = None,
    changepoint: Optional[dict] = None,
    collector: Optional[object] = None,
) -> Dict[str, float]:
    """
    Calculate average catchment storage trends.

    Uses simplified water balance (dS = P - Q) without ET estimation.
    Calculates storage amount corresponding to average discharge.

    Parameters
    ----------
    streamflow_data : pd.DataFrame
        Daily streamflow and climate data with columns:
            - water_year: int, water year
            - Q: float, daily discharge in mm/day
            - PPT: float, daily precipitation in mm/day
            - dowy: int, day of water year
    min_years : int, default 10
        Minimum number of valid years required

    Returns
    -------
    dict
        Dictionary of signature statistics with keys:
            - avg_storage_*: Average storage statistics (mm)

    Notes
    -----
    This calculation ignores evapotranspiration (ET), using only P - Q
    for the water balance. This simplification may overestimate storage
    in watersheds with significant ET losses.

    Reference:
        Peters, N.E., & Aulenbach, B.T. (2011). Water storage at
        the Panola Mountain Research Watershed.
    """
    # Validate required columns
    required_cols = ["water_year", "Q", "PPT", "dowy"]
    missing = [c for c in required_cols if c not in streamflow_data.columns]
    if missing:
        raise ValueError(f"Missing required columns: {missing}")

    df = streamflow_data
    years = df["water_year"].unique()

    # Return NAs if insufficient years
    if len(years) < min_years:
        return _empty_result()

    # Calculate storage for each year
    annual_storage = []

    for yr, year_data in df.groupby("water_year", sort=False):
        year_data = year_data.copy()  # defensive copy of the groupby slice (no in-place mutation follows)

        # Skip years with any remaining NAs (data should be pre-interpolated)
        if year_data["Q"].isna().any() or year_data["PPT"].isna().any():
            continue

        # Sort by day of water year
        year_data = year_data.sort_values("dowy")

        # Water balance: dS = P - Q
        dS = year_data["PPT"].values - year_data["Q"].values

        # Cumulative storage (relative to start of year)
        S = np.cumsum(dS)

        # Calculate mean Q for this year
        mean_Q = year_data["Q"].mean()

        # Find storage at mean Q using interpolation
        Q_values = year_data["Q"].values

        # Sort by Q for interpolation
        sort_idx = np.argsort(Q_values)
        Q_sorted = Q_values[sort_idx]
        S_sorted = S[sort_idx]

        # Remove NAs
        valid_mask = ~np.isnan(Q_sorted) & ~np.isnan(S_sorted)
        Q_valid = Q_sorted[valid_mask]
        S_valid = S_sorted[valid_mask]

        if len(Q_valid) < 10:
            continue

        # Average S values at duplicate Q points to match R's approx(ties=mean)
        # Without this, np.interp picks the last S value at tied Q points,
        # introducing noise that corrupts trend statistics
        Q_unique, inverse_idx = np.unique(Q_valid, return_inverse=True)
        S_means = np.zeros(len(Q_unique))
        counts = np.zeros(len(Q_unique))
        for idx_i, grp in enumerate(inverse_idx):
            S_means[grp] += S_valid[idx_i]
            counts[grp] += 1
        S_means /= counts

        if len(Q_unique) < 10:
            continue

        # Linear interpolation to find S at mean_Q
        try:
            avg_storage = np.interp(mean_Q, Q_unique, S_means)
        except Exception:
            avg_storage = np.nan

        if not np.isnan(avg_storage):
            annual_storage.append({
                "water_year": yr,
                "avg_storage": avg_storage
            })

    if len(annual_storage) < 5:
        return _empty_result()

    annual_df = pd.DataFrame(annual_storage)

    # Calculate trend statistics
    stats = generate_stats(
        annual_df,
        value_cols=["avg_storage"],
        year_col="water_year",
        trend_completeness=trend_completeness,
        decade_completeness=decade_completeness, min_values_for_stats=min_values_for_stats, changepoint=changepoint, collector=collector,
    )

    # Return the full generate_stats output (merged wholesale, as Julia does):
    # the previous explicit 8-key copy silently DROPPED the changepoint fields
    # (found 2026-08-24 by the Phase-1 schema gate: avg_storage was one of 3
    # bases missing its 8 _pettitt_* columns).
    return dict(stats)


def _empty_result() -> Dict[str, float]:
    """Return empty result with all NAs."""
    return {
        "avg_storage_senn_slp": np.nan,
        "avg_storage_linear_slp": np.nan,
        "avg_storage_spearman_rho": np.nan,
        "avg_storage_spearman_pval": np.nan,
        "avg_storage_mk_rho": np.nan,
        "avg_storage_mk_pval": np.nan,
        "avg_storage_mean": np.nan,
        "avg_storage_median": np.nan,
    }
